### PR TITLE
[SPARK-36147][SQL] Warn if less files visible after stats write in BasicWriteStatsTracker

### PR DESCRIPTION
**Pull request by tooptoop4**
### What changes were proposed in this pull request?

This log should at least be WARN not INFO (in org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala ) 
"Expected $numSubmittedFiles files, but only saw $numFiles."

### Why are the changes needed?

INFO logs don't indicate possible issue but WARN logs should

### Does this PR introduce any user-facing change?

Yes, Log level changed.

### How was this patch tested?

manual, trivial change

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@tooptoop4 please refer to https://spark.apache.org/contributing.html and make the PR description and title properly with a jira.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:169]: WDYT @steveloughran ?

Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:169]: This used to happen sometimes on code writing files to S3 where a cached HEAD-generated 404 was being returned even after the upload.
* Not going to be an issue on S3 any more, never was on GCS/Azure. 
* s3a magic committer creates marker files for this code
* Are there any FileOutputFormat's which don't write zero byte files if the client didn't write the output?

**Comment by steveloughran** [comment]
You are not going to see this on S3 any more from 404 cacheing.
That leaves the possibility that some FileOutputFormat doesn't save output (e.g. if there's no data). I don't know if that is the case. (I have vague memories of something (parquet?) adding the logic to skip reading 0 byte files for footers/data, but can't remember the details.)


It might be interesting here to actually add a counter of when this happens and return that as a stat in the map, at least that way you'll find out in the job reports without looking through logs.

Personally, I'd be interested in making the choice of stats tracker for hadoop FS insertion something pluggable. the s3a and abfs Input and output streams now collect stats on operations (see [iostatists](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/site/markdown/filesystem/iostatistics.md)). I'm collecting stats on IO done during task commit in the S3A and ABFS committers, aggregating them via the .json manifests and including in the final _SUCCESS report. That's handy, but it's not integrated with spark job reporting.

What I would really like to do would be collect stats of read and write streams used in a single task, report that, so you'd get details on #of throttle events, amout of data discarded in seeks, as the job went along. Which implies
* thread level collection of stats from all streams created in individual worker threads
* committers to retrieve and report 

Even without that, if I could plug in a new stats reporter, I'd be able to collect and report stats back from the manifest and s3a committers today.

**Comment by tooptoop4** [comment]
I did see this log printed on s3 yesterday

**Comment by HyukjinKwon** [comment]
@tooptoop4 can you file a JIRA, fill the PR description, and link it to the Pr title?

**Comment by tooptoop4** [comment]
@HyukjinKwon done

**Comment by HyukjinKwon** [comment]
can you keep the Pr description template? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

**Comment by tooptoop4** [comment]
> can you keep the Pr description template? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

@HyukjinKwon done

**Comment by steveloughran** [comment]
> I did see this log printed on s3 yesterday

ooh. s3a URL? Which committer and file output format? S3 no longer has 404 caching, and listings are fully consistent. So it isn't happening due to s3 state....something else is up. If it's not the output format skipping 0 byte files, something could be wrong.

can you grab the cloudstore jar, run its `storediag` command against your bucket  (ideally with -j for a JAR listing) & attach here, or email to me stevel@ at apache.org so I can see it: https://github.com/steveloughran/cloudstore

**Comment by tooptoop4** [comment]
s3a url, bucket is in one of asia pacific ones, it is accessed via a HTTP proxy and with STS.

some snippets of the log:
```
21-07-12 11:58:36 INFO org.apache.spark.sql.execution.datasources.FileScanRDD: Reading File path: s3a://xxxx/yyyy.csv, range: 0-1951, partition values: [empty row]
21-07-12 11:58:37 INFO org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: Code generated in 9.010979 ms
21-07-12 11:58:37 INFO org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: Code generated in 8.593598 ms
21-07-12 11:58:37 INFO org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: File Output Committer Algorithm version is 2
21-07-12 11:58:37 INFO org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: FileOutputCommitter skip cleanup _temporary folders under output directory:false, ignore cleanup failures: false
21-07-12 11:58:37 INFO org.apache.spark.sql.execution.datasources.SQLHadoopMapReduceCommitProtocol: Using user defined output committer class org.apache.parquet.hadoop.ParquetOutputCommitter
21-07-12 11:58:37 INFO org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: File Output Committer Algorithm version is 2
21-07-12 11:58:37 INFO org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: FileOutputCommitter skip cleanup _temporary folders under output directory:false, ignore cleanup failures: false
21-07-12 11:58:37 INFO org.apache.spark.sql.execution.datasources.SQLHadoopMapReduceCommitProtocol: Using output committer class org.apache.parquet.hadoop.ParquetOutputCommitter
21-07-12 11:58:37 INFO org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: Code generated in 6.470666 ms
21-07-12 11:58:37 INFO org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: Code generated in 4.507112 ms
21-07-12 11:58:37 INFO org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: Code generated in 13.859354 ms
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.codec.CodecConfig: Compression: SNAPPY
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.codec.CodecConfig: Compression: SNAPPY
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Parquet block size to 134217728
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Parquet page size to 1048576
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Parquet dictionary page size to 1048576
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Dictionary is on
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Validation is off
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Writer version is: PARQUET_1_0
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Maximum row group padding size is 8388608 bytes
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Page size checking is: estimated
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Min row count for page size check is: 100
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Max row count for page size check is: 10000
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Truncate length for column indexes is: 64
21-07-12 11:58:37 INFO org.apache.parquet.hadoop.ParquetOutputFormat: Page row count limit to 20000
21-07-12 11:58:37 INFO org.apache.hadoop.io.compress.CodecPool: Got brand-new compressor [.snappy]
21-07-12 11:58:38 INFO org.apache.spark.sql.execution.datasources.BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
21-07-12 11:58:38 INFO org.apache.spark.mapred.SparkHadoopMapRedUtil: No need to commit output of task because needsTaskCommit=false: attempt_xxxx836_0021_m_000000_617
```

can't see a fat jar of cloudstore and don't have access to hadoop installation atm
```
java -jar cloudstore-1.0.jar storediag
no main manifest attribute, in cloudstore-1.0.jar
```

**Comment by steveloughran** [comment]
First: You are using the the V2 FileOutputCommitt committer. 

This is dangerous as copies a task attempt's files during task commit. If for any reason a task is then reexecuted, those files will not be cleaned up. Unless you are 100% sure that the task files are the same and happy with intermingled file output from two tasks, it's not safe on *any* filesystem.

On S3 it is also really slow. V1 isn't safe their either, as it relies on atomic dir rename.

Switch to a zero rename committer, for both safety and performance. (I treat all support calls related to FOC on S3A as WONTFIX)

Second: Looking at the output:

```
21-07-12 11:58:38 INFO org.apache.spark.sql.execution.datasources.BasicWriteTaskStatsTracker: Expected 1 files, but only saw 0. This could be due to the output format not writing empty files, or files being not immediately visible in the filesystem.
21-07-12 11:58:38 INFO org.apache.spark.mapred.SparkHadoopMapRedUtil: No need to commit output of task because needsTaskCommit=false: attempt_xxxx836_0021_m_000000_617
````

Line 2 there says the entire task attempt dir doesn't exist, that's not just the file, it's the parent dir.

The parquet output writer *dit not write a file*. Either no records at all were created or close() was never called (remember: on S3 files don't get created until close() manifests them)

Summary: worry more about needsTaskCommit() == false and why that is surfacing.


That is independent of  committer; maybe S3 (on other stores the file would be empty (nothing written)/incomplete (not closed properly)), or task abort was called before the other probes (which deletes the dir...)

debugging strategies: 

* If you've set "spark.sql.maxConcurrentOutputFileWriters"  to >= 1, set it to 0. That guarantees is the non-current codepath. If the file is created then, but not in parallel runs, maybe something to do with the concurrent run not being complete.=
* switch to S3A committer and set org.apache.hadoop.fs.s3a.commit to log at debug. There's a lot of logging there.
* turn on all logging of org.apache.hadoop.fs.s3a to see what's happening
* use a different cluster store as the dest of the same job
* turn on s3 bucket logging and see what happened there. We're adding some really good auditing there, matching S3 REST calls to FS Calls, jobs IDs. Not yet shipping though.


[Side issue: @HyukjinKwon you ever thought of filing MAPREDUCE and PARQUET patches where you change those internal methods tagged @Private to @LimitedPrivate("spark") so those teams know what would break? It might give you some more stability.]

**Comment by tooptoop4** [comment]
can this log level change be merged?

**Comment by steveloughran** [comment]
@tooptoop4 I'm worried that this is possibly a serious problem. In which case actually listing what's missing might be good

**Comment by HyukjinKwon** [comment]
(I don't know much about this S3 context. I just left some comments to format the PR nicer. I'll defer to @steveloughran who has the best context here.)

**Comment by steveloughran** [comment]
just go with it as is; I'm looking at whether I could do a hadoop-3.3.1+ stats tracker which collects IOStatistics; rather than just report file length I want to aggregate all the stats we collect during stream reads/writes

**Comment by steveloughran** [comment]
(i.e. I'm planning to add some thread-context level collection of all the stream-specific stats; they'll the stats of the thread they were created in in their close() call (even if that's now a different thread); driver will get all this stuff to (somehow) report. At the very least they can print summary report somewhere

**Comment by tooptoop4** [comment]
can this log level change be merged? @steveloughran

**Comment by steveloughran** [comment]
+1 from me

**Comment by HyukjinKwon** [comment]
just to double check, is the change fine as is @steveloughran? I will just merge then.

**Comment by tooptoop4** [comment]
@HyukjinKwon he said +1

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
